src/adaptation/precond.jl

```diff
@@ -2,389 +2,227 @@
-NaiveVar(::Type{T}, ::Tuple{Int,Int}) where {T<:AbstractFloat} = NaiveVar(0, Vector{Matrix{T}}())
+struct UnitMassMatrix{T<:AbstractFloat} <: MassMatrixAdaptor end
 
-function add_sample!(nc::NaiveVar, s::AbstractVecOrMat)
-    nc.n += 1
-    push!(nc.S, s)
-end
+Base.show(io::IO, ::UnitMassMatrix) = print(io, "UnitMassMatrix")
 
-function reset!(nc::NaiveVar{T}) where {T}
-    nc.n = 0
-    nc.S = T()
-end
+UnitMassMatrix(::Type{T}=Float64) where {T} = UnitMassMatrix{T}()
 
-function get_var(nc::NaiveVar)
-    @assert nc.n >= 2 "Cannot get variance with only one sample"
-    return Statistics.var(nc.S)
-end
+Base.string(::UnitMassMatrix) = "I"
 
-# Ref： https://github.com/stan-dev/math/blob/develop/stan/math/prim/mat/fun/welford_var_estimator.hpp
-mutable struct WelfordVar{T<:AbstractVecOrMat{<:AbstractFloat}} <: VarEstimator{T}
-    n :: Int
-    μ :: T
-    M :: T
-    # TODO: implement temporary `δ` as `WelfordCov`
-end
+Base.resize!(pc::UnitMassMatrix, θ::AbstractVecOrMat) = nothing
 
-WelfordVar(::Type{T}, sz::Union{Tuple{Int},Tuple{Int,Int}}) where {T<:AbstractFloat} = WelfordVar(0, zeros(T, sz), zeros(T, sz))
+reset!(::UnitMassMatrix) = nothing
 
-function reset!(wv::WelfordVar{<:AbstractVecOrMat{T}}) where {T<:AbstractFloat}
-    wv.n = 0
-    wv.μ .= zero(T)
-    wv.M .= zero(T)
-end
+getM⁻¹(::UnitMassMatrix{T}) where {T} = LinearAlgebra.UniformScaling{T}(one(T))
 
-function add_sample!(wv::WelfordVar, s::AbstractVecOrMat)
-    wv.n += 1
-    @unpack μ, M, n = wv
-    for i in eachindex(s)
-        δ = s[i] - μ[i]
-        μ[i] += δ / n
-        M[i] += δ * (s[i] - μ[i])
-    end
-end
+adapt!(
+    ::UnitMassMatrix,
+    ::AbstractVecOrMat{<:AbstractFloat},
+    ::AbstractScalarOrVec{<:AbstractFloat},
+    is_update::Bool=true
+) = nothing
 
-# https://github.com/stan-dev/stan/blob/develop/src/stan/mcmc/var_adaptation.hpp
-function get_var(wv::WelfordVar{<:AbstractVecOrMat{T}}) where {T<:AbstractFloat}
-    n, M = T(wv.n), wv.M
-    @assert n >= 2 "Cannot get covariance with only one sample"
-    return (n / ((n + 5) * (n - 1))) .* M .+ T(1e-3) * (5 / (n + 5))
-end
+## Diagonal mass matrix adaptor
 
-## Covariance estimator
+abstract type VarEstimator{T} <: MassMatrixAdaptor end
 
-abstract type CovEstimator{T} end
+getM⁻¹(ve::VarEstimator) = ve.var
 
-# NOTE: This naive covariance estimator is used only in testing.
-mutable struct NaiveCov{T<:AbstractVector{<:AbstractVector{<:AbstractFloat}}} <: CovEstimator{T}
+Base.string(ve::VarEstimator) = string(getM⁻¹(ve))
+
+function adapt!(
+    ve::VarEstimator,
+    θ::AbstractVecOrMat{<:AbstractFloat},
+    α::AbstractScalarOrVec{<:AbstractFloat},
+    is_update::Bool=true
+)
+    resize!(ve, θ)
+    push!(ve, θ)
+    if ve.n >= ve.n_min && is_update
+        ve.var .= getest(ve)
+    end
+end
+
+# NOTE: this naive variance estimator is used only in testing
+mutable struct NaiveVar{E<:AbstractVecOrMat{<:AbstractFloat},T<:AbstractVector{E}} <: VarEstimator{T}
```

Comment:
Does it have to be mutable?

Reply:
I saw you increase `n` below, but `n` just seems to be the length of the vector `S`. I'd suggest just calling `length(S)` if needed and not saving `n`.

Reply:
I'd also suggest making the element type a type parameter rather (or in addition to) `E`:
```suggestion
struct NaiveVar{S<:AbstractFloat,T<:AbstractVector{<:AbstractVecOrMat{S}}} <: VarEstimator{T}
```
Maybe it would also be more important to know the element type in `VarEstimator`? Then it should rather be `VarEstimator{S}`.

Reply:
> I saw you increase n below, but n just seems to be the length of the vector S. I'd suggest just calling length(S) if needed and not saving n.

Yes it's only for naive var/covar estimator though. For Welford, `n` is not accessible. In fact naive estimators are not used but only for test purposes. But I can do your suggestion anyway.